### PR TITLE
Trying out logprobs and top logprobs for testing rather than logits.

### PR DESCRIPTION
**Pull request by Manan17**
## Summary

Just testing out logprobs as mentioned in #742 

## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: test/convergence/bf16/test_mini_models.py [hunk 1]
```diff
@@ -1189,13 +1189,32 @@ def test_mini_model(
...
+        
         assert_verbose_allclose(
-            expected_output["logits"],
-            actual_output["logits"],
+            expected_logprobs,
+            actual_logprobs,
             atol=logits_atol,
-            rtol=logits_rtol,
+            rtol=logits_rtol
         )
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py]: We don't have to compare all logprobs

File: test/convergence/bf16/test_mini_models.py [hunk 2]
```diff
@@ -1189,13 +1189,32 @@ def test_mini_model(
...
+        exp_topk_vals, _ = torch.topk(expected_logprobs, k, dim=-1)
+        act_topk_vals, _ = torch.topk(actual_logprobs, k, dim=-1)
 
+        # Compare top-k logprobs with tolerance (ignoring order)
+        max_diff = torch.max(torch.abs(exp_topk_vals - act_topk_vals)).item()
+        print(f"Top-{k} logprobs max diff: {max_diff:.6f}")
+        assert torch.all(torch.abs(exp_topk_vals - act_topk_vals) < (logits_atol + logits_rtol * torch.abs(exp_topk_vals))), (
+            f"Top-{k} logprobs are not all close (atol={logits_atol}). "
+            f"Max diff: {max_diff:.6f}"
+        )
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py]: Make it a test util

File: test/convergence/bf16/test_mini_models.py [hunk 3]
```diff
@@ -896,7 +896,7 @@ def run_mini_model(
             torch.bfloat16,
             1e-3,
             1e-2,
-            1,  # 1e-1
+            1e-1,  # 1e-1
             1e-1,  # 1e-2
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py:908]: After removing all logprobs comparison, we can try setting it lower.
sglang only has atol and sets it to 5e-2 ([decode_tolerance](https://github.com/search?q=repo%3Asgl-project/sglang%20decode_tolerance&type=code))
verl sets [`(atol, rtol) = (1e-2, 1e-5)`](https://github.com/volcengine/verl/blob/0e127b208b7ec89bbd3647668973a35fe5d81390/tests/models/test_transformer.py#L134), but it's mean of all logprobs not topk

Comment (Manan17) [reply test/convergence/bf16/test_mini_models.py:908]: Does not work with lower tolerance.
For gemma3, it passes when atol=1e-1 and rtol=1

Comment (Manan17) [reply test/convergence/bf16/test_mini_models.py:908]: I tested this out with fp32, it fails for most of the models where old logic for checking the logits is passing.

File: test/utils.py [hunk 4]
```diff
@@ -57,6 +57,19 @@ def set_seed(seed=42):
...
 
+def check_logprobs(tensor1, tensor2, rtol=1e-05, atol=1e-08, k=5, max_print=5):
+    act_topk_vals, _ = torch.topk(tensor1, k, dim=-1)
+    exp_topk_vals, _ = torch.topk(tensor2, k, dim=-1)
+
+    # Compare top-k logprobs with tolerance (ignoring order)
+    max_diff = torch.max(torch.abs(exp_topk_vals - act_topk_vals)).item()
+    print(f"Top-{k} logprobs max diff: {max_diff:.6f}")
+    assert torch.all(torch.abs(exp_topk_vals - act_topk_vals) < (rtol + rtol * torch.abs(exp_topk_vals))), (
+        f"Top-{k} logprobs are not all close (tolerance={rtol + rtol * torch.abs(exp_topk_vals)}). "
```
Comment (Tcc0403) [inline test/utils.py:68]:
```suggestion
    assert torch.all(torch.abs(exp_topk_vals - act_topk_vals) < (atol + rtol * torch.abs(exp_topk_vals))), (
        f"Top-{k} logprobs are not all close (tolerance={atol + rtol * torch.abs(exp_topk_vals)}). "
```

File: test/convergence/bf16/test_mini_models.py (hunk 3 above)
Comment (Tcc0403) [reply test/convergence/bf16/test_mini_models.py:908]: Since we are comparing values in log-space, the total tolerance here is actually relative tolerance.

File: test/convergence/bf16/test_mini_models.py [hunk 5]
```diff
@@ -1189,13 +1190,10 @@ def test_mini_model(
...
     if expected_output["logits"] is not None and actual_output["logits"] is not None:
-        assert_verbose_allclose(
-            expected_output["logits"],
-            actual_output["logits"],
-            atol=logits_atol,
-            rtol=logits_rtol,
-        )
-
+        actual_logprobs = torch.nn.functional.log_softmax(actual_output["logits"], dim=-1)
+        expected_logprobs = torch.nn.functional.log_softmax(expected_output["logits"], dim=-1)
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py]: Make `log_softmax()` and `topk()` a util function, so we can call it in `run_mini_model()` to avoid storing all logits

File: test/convergence/bf16/test_mini_models.py [hunk 6]
```diff
@@ -1189,13 +1190,10 @@ def test_mini_model(
...
+        check_logprobs(actual_logprobs,expected_logprobs, atol=logits_atol,rtol=logits_rtol)
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py]: Assuming we have top logprobs calculated and stored, we only need to call 
`assert_verbose_allclose(actual_top_logprobs, expected_top_logprobs, atol=logprob_atol, rtol=logprob_rtol)`.

File: test/convergence/bf16/test_mini_models.py (hunk 3 above)
Comment (Manan17) [reply test/convergence/bf16/test_mini_models.py:908]: Can we just check the rtol?
like: tolerance = rtol * torch.abs(tensor2)

**Comment by Manan17** [comment]
@Tcc0403 Can you have a look at the changes, I have tested it. 
Let me know what you think, I will update it for the multimodal tests as well. 
What should be done for test_mini_models_with_logits?

File: test/convergence/bf16/test_mini_models.py (hunk 3 above)
Comment (Tcc0403) [reply test/convergence/bf16/test_mini_models.py:908]: absolute diff for two logprobs (logA - logB) = relative diff for two probs (A / B), which means the whole tolerance (`atol + rtol * torch.abs(expected)`) should be the maximum relative diff we can accept. 

I think that's also why sglang only has a single tolerance in their test.

**Comment by Tcc0403** [comment]
> What should be done for test_mini_models_with_logits

check logprobs as well for consistency

I'm planning to rewrite convergence tests so just ignore namings for now.

File: test/utils.py [hunk 7]
```diff
@@ -56,6 +56,13 @@ def set_seed(seed=42):
     # Python hash seed
     os.environ["PYTHONHASHSEED"] = str(seed)
 
+def get_logprobs(tensor):
+    return torch.nn.functional.log_softmax(tensor, dim=-1)
+
+def get_topk(tensor, k=5):
+    topk = torch.topk(tensor, k, dim=-1)
+    return topk
```
Comment (Tcc0403) [inline test/utils.py]: add decorator [`@torch.no_grad`](https://docs.pytorch.org/docs/stable/generated/torch.no_grad.html)

**Comment by Manan17** [comment]
> > What should be done for test_mini_models_with_logits
> 
> check logprobs as well for consistency
> 
> I'm planning to rewrite convergence tests so just ignore namings for now.

Gotcha!
I tried testing with mean logprobs as well. 
The tests pass with lower tolerance values. Verl has set atol=1e-2 and rtol=1e-5, which works for us as well in bf16.

**Comment by Tcc0403** [comment]
> I tried testing with mean logprobs as well.
> The tests pass with lower tolerance values. Verl has set atol=1e-2 and rtol=1e-5, which works for us as well in bf16.

What mean logprobs do you pick? I checked verl impl, they pick [per-token logprobs for the given labels](https://github.com/volcengine/verl/blob/16662ceff4c90c0893dd2d249965bed94f2a5e1a/verl/utils/torch_functional.py#L56)

File: test/utils.py [hunk 8]
```diff
@@ -56,6 +56,15 @@ def set_seed(seed=42):
     # Python hash seed
     os.environ["PYTHONHASHSEED"] = str(seed)
 
+@torch.no_grad
+def get_logprobs(tensor):
+    return torch.nn.functional.log_softmax(tensor, dim=-1)
```
Comment (Tcc0403) [inline test/utils.py]: For numerical stability if tensor is bf16
```suggestion
@torch.no_grad
def get_logprobs(tensor):
    return torch.nn.functional.log_softmax(tensor, dim=-1, dtype=torch.float32) # fp32 for stability
```

**Comment by Manan17** [comment]
I tried top 20 logprobs and it was able to pass tests for all the models! @Tcc0403

**Comment by Manan17** [comment]
The tolerance for gemma3 multimodal model had to be set high as it does not pass the tests for loss and topk_logprobs. 
The atol and rtol set it 1e-1.

**Comment by Tcc0403** [comment]
> The tolerance for gemma3 multimodal model had to be set high as it does not pass the tests for loss and topk_logprobs.
> The atol and rtol set it 1e-1.

Yeah, I think we can compromise with 1e-1 before further investigation in numerical issue. Just make them all green first unless there's an obvious mismatch.

**Review by shimizust — APPROVED** [review]
Thanks for making these changes!
